ath10k: enable encapsulation offload by default

tiagogaspar8 opened the pull request:
Enable ath10k offload by default. This improves wireless performance without requiring user configuration.

Reports suggest that on some ath10k devices there can be about 10% improvement on tx throughput.

Signed-off-by: Tiago Gaspar <tiagogaspar8@gmail.com>

rsa9000 commented:
+1

My and other community members tests show [1] that offloading is stable enough to be enabled by default.

1. https://www.mail-archive.com/openwrt-devel@lists.openwrt.org/msg62013.html

Ansuel commented:
@chunkeey if you have nothing against this i would merge this. I also testes this for prolonged time (at times) and with plain ath10k this works correctly.

PolynomialDivision commented:
Also tested with 802.11s mesh.

tiagogaspar8 commented:
Hi all,

I just update the PR, the module name was wrong but it is now correct

dhewg commented:
Maybe I'm blind, but that doesn't seem to do anything?
If it's supposed to add the param to /etc/modules.d/ath10k you probably meant:
```
  AUTOLOAD:=$(call AutoProbe,ath10k_core ath10k_pci)
  MODPARAMS.ath10k_core:=frame_mode=2
```

That seems to do the trick:
```
$ cat /etc/modules.d/ath10k
ath10k_core frame_mode=2
ath10k_pci
```

And:
```
$ cat /sys/module/ath10k_core/parameters/frame_mode
2
```

How did you verify this at runtime?

johnnyslee commented:
May I suggest adding a Kconfig defaults to the desired value instead of hard-coding it?

rsa9000 commented:
@tiagogaspar8, I personally like your solution with specifying the module parameter on load. It is compact and simplifies the backports rebase. But accounting the @dhewg comment that this approach requires a bigger package rework, I am tend to wonder, maybe we should just patch the driver and change the defaults there?

rsa9000 commented:
> May I suggest adding a Kconfig defaults to the desired value instead of hard-coding it?

Yep, this is looks possible. But why do we need it?

johnnyslee commented:
> > May I suggest adding a Kconfig defaults to the desired value instead of hard-coding it?
> 
> Yep, this is looks possible. But why do we need it?

At first I thought we might need a way to change the value if the new mode didn't work for a specific device, but on second thought, a Kconfig in the package wouldn't help in that case. Including a custom config file should be sufficient if that's the case. So, forget it. We don't need a Kconfig.

tapper82 commented:
If this is not on by default then you need to let people know how to use it. Plus make it simple to toggel. Has any one made a wiki post about what this is and what it does? Why would some one want to use it and what can it do for them?

robimarko commented:
@tapper82 There is nothing people need to know, it's set and forget.
If they want to disable it, just change the config file that's generated in `/etc/modules.d`

The PR author explained the benefits, you get better throughput as the driver will offload encapsulating TX packets to the firmware instead of doing it via mac80211.

robimarko commented:
@tapper82 The best and easiest thing would be to link the patch series in the commit description:
https://patchwork.kernel.org/project/linux-wireless/cover/20220516032519.29831-1-ryazanov.s.a@gmail.com/

It's quite detailed and everything is there, which will make it easy to figure everything out.
Wiki is quite useless for these kinds of things as googling is gonna get you way faster to stuff like this instead of the wiki.

Ansuel commented:
My 2 cent on the topic... Considering the perf increase this option should be enabled by default and any related problem should be fixed and reported to ath10k. If some problem comes we ask for logs or repro steps and we provide workaround on how to revert this. 

I see an additional config flag a bit too much.... doable but again "In a perfect world" someone would use frame_mode 2 and don't care about other...

Also this is really about encapsulation offload... so it's really a thing of it does work with the configuration or it doesn't work at all and the firmware doesn't transmit packet...

tiagogaspar8 commented:
So, I don't believe there is a need for a different package of an option to enable this separately, because as ansuel said, this is proven tech and people can always report bugs, if there are any.
On another note @dhewg, you were right, I made a mistake in my testing, it is now corrected and fully tested, this time the right way 😉 , I also followed robimarko's suggestion and added the link to the patch in my commit so people are correctly informed 😄

Ansuel commented:
@tiagogaspar8 hope you are not hating me... wonder if you can also add some info like

`This set the frame_mode to ethernet mode (2). This can be set by passing a different value to the frame_mode value on kmod insmod`

tiagogaspar8 committed:
ath10k: enable encapsulation offload by default

Enable ath10k offload by default. This improves wireless performance
without requiring user configuration.
This adds ath10k_core to the AUTOLOAD section so that the frame_mode
paramter can be added to /etc/modules.d and passed to the driver.
The frame_mode 2 enables ethernet mode on the firmware/driver.
This parameter is set by passing a different value to the frame_mode
value on kmod insmod.
Link to the original patchset:
https://patchwork.kernel.org/project/linux-wireless/cover/20220516032519.29831-1-ryazanov.s.a@gmail.com/

Signed-off-by: Tiago Gaspar <tiagogaspar8@gmail.com>

chunkeey commented:
> @chunkeey if you have nothing against this i would merge this. I also testes this for prolonged time (at times) and with plain ath10k this works correctly.

Sorry, I'm mostly away during the week. In this case, it looks like it helped to iron out a bug. No, go right ahead with your plan.

Ansuel commented:
Thanks, merged with 40c2cd8bd0f7e7b6553d47c11dacb48acc533863

tapper82 commented:
I see that people are asking the same things on the PR for the backport.

Ansuel commented:
@tapper82 the thing is that you don't need to do anything. Just report if you encounter bugs, that's all.
i don't think that will be backported to 22.03 since it's still not widely tested and we need to give stability a priority

Ansuel commented:
https://github.com/openwrt/openwrt/pull/10095#issuecomment-1162016812 ....
It's always on.

tapper82 commented:
So I can't use this with ath10k CT. What's the point then? OpenWrt ships with the CT build of ath10k.

Ansuel commented:
It's not our fault... go complain on ath10k-ct for not supporting this feature... This finally got merged upstream so we are setting it on by default on ath10k. Community or someone can just build its image and use ath10k instead of ath10k-ct.
